t/core/ctx.t

```diff
@@ -155,3 +155,264 @@ GET /t?a=aaa
+                }]],
+                [[{
+                    "node": {
+                        "value": {
+                            "methods": [
+                                "GET"
+                            ],
+                            "uri": "/hello",
+                            "plugins": {
+                                "serverless-pre-function": {
+                                    "phase": "access",
+                                    "functions" : ["return function() ngx.log(ngx.ERR, \"route_id: \", ngx.ctx.api_ctx.route_id) end"]
+                                }
+                            },
+                            "upstream": {
+                                "type": "roundrobin",
+                                "nodes": {
+                                    "127.0.0.1:1980": 1
+                                }
+                            }
+                        },
+                        "key": "/apisix/routes/1"
+                    },
+                    "action": "set"
+                }]]
+                )
+
+            ngx.status = code
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: `url` exist and `route_id` is not empty
+--- request
+GET /hello
+--- response_body
+hello world
+--- error_log
+route_id: 1
+
+
+
+=== TEST 9: `url` does not exist and  `route_id` is empty
+--- request
+GET /hello123
+--- error_code: 404
+--- no_error_log
+[error]
+
+
+
+=== TEST 10: create a service and `service_id` is 1
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/services/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                    "desc": "new_service"
+                }]],
+                [[{
+                    "node": {
+                        "value": {
+                            "desc": "new_service"
+                        },
+                        "key": "/apisix/services/1"
+                    },
+                    "action": "set"
+                }]]
+                )
+
+            ngx.status = code
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 11: the route object not bind any service object
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "methods": ["GET"],
+                    "plugins": {
+                        "serverless-pre-function": {
+                            "phase": "access",
+                            "functions" : ["return function() if ngx.ctx.api_ctx.service_id then ngx.log(ngx.ERR, \"service_id: \", ngx.ctx.api_ctx.service_id) end end"]
```

Comment:
`ngx.ctx.api_ctx.service_id` -> `ngx.ctx.api_ctx.var["service_id"]` ?

Reply:
we should print the `service_id` always, and check the output value in error log.

`ngx.log(ngx.INFO, "service_id: ", ngx.ctx.api_ctx.var["service_id"] or "empty id")`

Reply:
thank you, I got it.